### ISSUE TITLE
Implement DbProviderFactory

### ISSUE THREAD

**Issue by PromontoryProtean** [aspnet/Microsoft.Data.Sqlite#115]
Is there any plan to add DbProviderFactory support to this package and/or vnext in general? I am attempting to write a DNX based app that needs interchangeable SQL Server and Sqlite support, and I am using Dapper for data access. Is there any general guidance that you can offer for this type of scenario?

**Comment by bricelam** [comment]
The [dotnet/corefx](https://github.com/dotnet/corefx) may be a better place to file this. If it gets added to the System.Data.Common contract, we'll implement it.

**Comment by natemcmaster** [comment]
Corefx includes [DbProviderFactory abstract class](https://github.com/dotnet/corefx/blob/master/src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs) in System.Data.Common. Does this mean it's in the contract?

**Comment by bricelam** [comment]
Interesting... I wonder when that happened. :smile: We should implement it.

**Comment by divega** [comment]
Yes. What is intentionally missing from `System.Data.Common` is `DbProviderFactories` class and the config dependent registration mechanism it supported, but `DbProviderFactory` is still there and I think we should support it, especially if we can get feedback from the community on what makes it important.

@PromontoryProtean can you elaborate on how your application and/or Dapper is using the provider factories?

**Comment by PromontoryProtean** [comment]
@divega, I'll do my best to explain. My company maintains a proprietary CMS that we are looking to migrate to vnext. It currently uses Sqlite for small installations and SQL Server for larger customers that get heavy traffic. We are focused heavily on performance so we wrote our data layer using raw SQL and Dapper, use DbProviderFactory to write DB agnostic code (with SQL dialect adjustments where needed) and define the SQL engine via web.config. When using Sqlite we use the System.Data.Sqlite provider and wire it up with the following sections in web.config.

    <connectionStrings>
        <add name="CmsConnection" providerName="System.Data.SQLite" connectionString="Data Source=|DataDirectory|\Sqlite\Cms.db;Version=3;BinaryGUID=False;datetimekind=Utc;" />
    </connectionStrings>

    <system.data>
        <DbProviderFactories>
            <remove invariant="System.Data.SQLite" />
            <add name="SQLite Data Provider" invariant="System.Data.SQLite" description=".Net Framework Data Provider for SQLite" type="System.Data.SQLite.SQLiteFactory, System.Data.SQLite" />
        </DbProviderFactories>
    </system.data>

While exploring the vnext code here on GitHub I discovered the Microsoft.Data.Sqlite repository. I was thinking I could use this as a replacement for System.Data.Sqlite, so I installed it via Nuget and began trying to wire it up. I got far enough along to get the following error:

`ArgumentException: Unable to find the requested .Net Framework Data Provider. It may not be installed.`

I realized I hadn't defined DbProviderFactories in any type of config and I didn't know what the vnext equivalent to it was, so I started looking through the source here. I searched for anything related to DbProviderFactory and saw that nothing had been implemented, and that led to my post.

I get the impression that this repo is mainly intended to be used with Entity Framework 7, but it would be great to open this up for use in different scenarios if possible. Are there any short or long term plans to enable a scenario such as what I outlined above?
